lib/core/dns-resolver.js

```diff
@@ -0,0 +1,443 @@
+
+    if (cached) {
+      this.stats.cache++
+    }
+
+    if (!cached) {
+      const pending = this._pending[hostname]
+      if (pending) {
+        this.stats.cache++
+        cached = await pending
+      } else {
+        const newPromise = this.queryAndCache(hostname)
+        this._pending[hostname] = newPromise
+        this.stats.query++
+        try {
+          cached = await newPromise
+        } finally {
+          delete this._pending[hostname]
+        }
+      }
+    }
+
+    return cached
+  }
+
+  async _resolve (hostname) {
+    // ANY is unsafe as it doesn't trigger new queries in the underlying server.
+    const entries = await Promise.allSettled([
+      ignoreNoResultErrors(this._resolve4(hostname, ttl)),
+      ignoreNoResultErrors(this._resolve6(hostname, ttl))
+    ])
+
+    if (entries[0].status === 'rejected' && entries[1].status === 'rejected') {
+      throw entries[0].reason
+    }
+
+    const A = entries[0].status === 'fulfilled' ? entries[0].value : []
+    const AAAA = entries[1].status === 'fulfilled' ? entries[1].value : []
+
+    let aTtl = 0
+    let aaaaTtl = 0
+    let cacheTtl = 0
+
+    const now = Date.now()
+
+    for (const entry of A) {
+      entry.family = 4
+      entry.expires = now + entry.ttl * 1000
+
+      aTtl = Math.max(aTtl, entry.ttl)
+    }
+
+    for (const entry of AAAA) {
+      entry.family = 6
+      entry.expires = now + entry.ttl * 1000
+
+      aaaaTtl = Math.max(aaaaTtl, entry.ttl)
+    }
+
+    if (A.length > 0) {
+      if (AAAA.length > 0) {
+        cacheTtl = Math.min(aTtl, aaaaTtl)
+      } else {
+        cacheTtl = aTtl
+      }
+    } else {
+      cacheTtl = aaaaTtl
+    }
+
+    return {
+      entries: [...AAAA, ...A],
+      cacheTtl
+    }
+  }
+
+  async _lookup (hostname) {
+    try {
+      const entries = await Promise.allSettled([
+        // Passing {all: true} doesn't return all IPv4 and IPv6 entries.
+        // See https://github.com/szmarczak/cacheable-lookup/issues/42
+        ignoreNoResultErrors(this._dnsLookup(hostname, all4)),
+        ignoreNoResultErrors(this._dnsLookup(hostname, all6))
+      ])
+
+      const A = entries[0].status === 'fulfilled' ? entries[0].value : []
+      const AAAA = entries[1].status === 'fulfilled' ? entries[1].value : []
+
+      return {
+        entries: [...AAAA, ...A],
+        cacheTtl: 0
+      }
+    } catch {
+      return {
+        entries: [],
+        cacheTtl: 0
+      }
+    }
+  }
+
+  async _set (hostname, data, cacheTtl) {
+    if (this.maxTtl > 0 && cacheTtl > 0) {
+      cacheTtl = Math.min(cacheTtl, this.maxTtl) * 1000
+      data[kExpires] = Date.now() + cacheTtl
+
+      try {
+        await this._cache.set(hostname, data, cacheTtl)
+      } catch (error) {
+        this.lookupAsync = async () => {
+          const cacheError = new Error(
+            'Cache Error. Please recreate the DNSResolver instance.'
+          )
+          cacheError.cause = error
+
+          throw cacheError
```

Comment:
I don't think this kind of error handling is correct. Error situations should auto heal when possible

Reply:
Here it is because it's possible to specify a custom cache object, if the set method throws it is considered as a fatal and unrecoverable error.

Using the default cache which is a simple Map() this would not happen.

Also happy to remove any form of custom caching, that's was part of the copied module.